test/core/client_channel/client_channel_test.cc

```diff
@@ -12,79 +11,268 @@
+      auto handler = call_destination_->PopHandler();
+      if (handler.has_value()) return std::move(*handler);
+      return Pending();
+    });
+  }
+
+  void QueueNameResolutionResult(Resolver::Result result) {
+    if (resolver_ != nullptr) {
+      resolver_->QueueNameResolutionResult(std::move(result));
+    } else {
+      early_resolver_results_.push(std::move(result));
+    }
+  }
+
+  Resolver::Result MakeSuccessfulResolutionResult(
+      absl::string_view endpoint_address) {
+    Resolver::Result result;
+    grpc_resolved_address address;
+    CHECK(grpc_parse_uri(URI::Parse(endpoint_address).value(), &address));
+    result.addresses = EndpointAddressesList({EndpointAddresses{address, {}}});
+    return result;
+  }
+
+ private:
+  class TestConnector final : public SubchannelConnector {
+   public:
+    void Connect(const Args&, Result*, grpc_closure* notify) override {
+      CHECK_EQ(notify_, nullptr);
+      notify_ = notify;
+    }
+
+    void Shutdown(grpc_error_handle error) override {
+      if (notify_ != nullptr) ExecCtx::Run(DEBUG_LOCATION, notify_, error);
+    }
+
+   private:
+    grpc_closure* notify_ = nullptr;
+  };
+
+  class TestClientChannelFactory final : public ClientChannelFactory {
+   public:
+    RefCountedPtr<Subchannel> CreateSubchannel(
+        const grpc_resolved_address& address,
+        const ChannelArgs& args) override {
+      gpr_log(GPR_INFO, "CreateSubchannel: args=%s", args.ToString().c_str());
+      return Subchannel::Create(MakeOrphanable<TestConnector>(), address, args);
+    }
+  };
+
+  class TestCallDestination final : public UnstartedCallDestination {
+   public:
+    void StartCall(UnstartedCallHandler unstarted_call_handler) override {
+      handlers_.push(
+          unstarted_call_handler.V2HackToStartCallWithoutACallFilterStack());
+    }
+
+    absl::optional<CallHandler> PopHandler() {
+      if (handlers_.empty()) return absl::nullopt;
+      auto handler = std::move(handlers_.front());
+      handlers_.pop();
+      return handler;
+    }
+
+    void Orphaned() override {}
+
+   private:
+    std::queue<CallHandler> handlers_;
+  };
 
-TEST(MakeSubchannelArgs, ArgsFromChannelTrumpPerAddressArgs) {
-  ChannelArgs args = ClientChannelFilter::MakeSubchannelArgs(
-      ChannelArgs().Set("foo", 1), ChannelArgs().Set("foo", 2), nullptr,
-      "foo.example.com");
-  EXPECT_EQ(args.GetInt("foo"), 1);
+  class TestCallDestinationFactory final
+      : public ClientChannel::CallDestinationFactory {
+   public:
+    explicit TestCallDestinationFactory(ClientChannelTest* test)
+        : test_(test) {}
+
+    RefCountedPtr<UnstartedCallDestination> CreateCallDestination(
+        ClientChannel::PickerObservable picker) override {
+      CHECK(!test_->picker_.has_value());
+      test_->picker_ = std::move(picker);
+      return test_->call_destination_;
+    }
+
+   private:
+    ClientChannelTest* const test_;
+  };
+
+  class TestResolver final : public Resolver {
```

Comment:
The fake resolver has so many quirks at this point that I'd rather not use it - especially with the simplicity of this one.

Reply:
I'll leave it up to you, but the fake resolver got a lot simpler in #34887, and I suspect it would be easy to use here.  My concern here is mainly that the argument "the existing thing is too complex, so let's build a new one" will ultimately lead to something like https://xkcd.com/927/. :)  If there are things that can be simplified about the existing implementation, I'd rather we do that instead of having multiple implementations.